### PR TITLE
[Merged by Bors] - feat: coalgebra isomorphisms

### PR DESCRIPTION
**Pull request by 101damnations**
Isomorphisms of coalgebras. Mimics `Mathlib.Algebra.Module.Equiv`.

---

- [x] depends on: #11961 
- [x] depends on: #11966 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Coalgebra/Equiv.lean [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+@[simp, norm_cast]
+theorem toCoalgHom_inj {e₁ e₂ : A ≃ₗc[R] B} : (↑e₁ : A →ₗc[R] B) = e₂ ↔ e₁ = e₂ :=
+  toCoalgHom_injective.eq_iff
+
+@[simp]
+theorem coe_mk {f h h₀ h₁ h₂ h₃ h₄ h₅} :
+    (⟨⟨⟨⟨f, h⟩, h₀⟩, h₁, h₂⟩, h₃, h₄, h₅⟩ : A ≃ₗc[R] B) = f := rfl
+
+theorem coe_injective : @Function.Injective (A ≃ₗc[R] B) (A → B) CoeFun.coe :=
```
Comment (eric-wieser) [inline Mathlib/RingTheory/Coalgebra/Equiv.lean:115]: This is not the right statement; it should be about `DFunlike.coe`.

I think you can just remove this lemma entirely?

**Review by eric-wieser — APPROVED** [review]
bors d+

It pains me to see us writing so much boilerplate for every algebraic morphism, but until automation appears I think we have to keep going.

Just some minor comments.

File: Mathlib/RingTheory/Coalgebra/Equiv.lean [hunk 2]
```diff
@@ -0,0 +1,253 @@
+/-
+Copyright (c) 2024 Amelia Livingston. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nathaniel Thomas, Jeremy Avigad, Johannes Hölzl, Mario Carneiro, Anne Baanen,
+  Frédéric Dupuis, Heather Macbeth, Amelia Livingston
```
[the remark this answers is not in the document]
Comment (101damnations) [reply Mathlib/RingTheory/Coalgebra/Equiv.lean:5]: Thank you! Done

**Pull request closed without merging**